test/studies/ml/lib/Tensor.chpl

```diff
@@ -0,0 +1,925 @@
+        }
+        return grad;
+    }
+
+    // Compute the gradient of a loss with respect to an input
+    proc correlateWeight(const ref filter: Tensor(2), pIn: 2*int, pOut: 2*int, stride: int = 1, padding: int = 0) {
+        const (m,n) = pIn;
+        const (i,j) = pOut;
+        const diff = (m - (stride * i - padding), n - (stride * j - padding));
+        const (dx,dy) = diff;
+        const (kh,kw) = filter.shape;
+        if dx >= 0 && dy >= 0 && dx < kh && dy < kw then
+            return filter[diff];
+        return 0.0;
+    }
+
+    // Compute the index of the gradient of a loss with respect to an input. Probably the most important function here.
+    proc correlateWeightIdx(filterShape: 2*int, pIn: 2*int, pOut: 2*int, stride: int = 1, padding: int = 0) {
+        const (m,n) = pIn;
+        const (i,j) = pOut;
+        const (dx,dy) = (m - (stride * i - padding), n - (stride * j - padding));
+        const (kh,kw) = filterShape;
+        if dx >= 0 && dy >= 0 && dx < kh && dy < kw then
+            return (dx,dy);
+        return (-1,-1);
+    }
+    
+    // Softmax but returns the sum of the exponentials and the exponentials themselves
+    proc softmaxParts(t: Tensor(?rank)) {
+        const m = max reduce t.data;
+        var y = t;
+        y.data -= m;
+        foreach i in y.data.domain {
+            y.data[i] = Math.exp(y.data[i]);
+        }
+        const sum = + reduce y.data;
+        
+        return (y,sum,y / sum);
+    }
+
+    // Softmax function
+    proc softmax(t: Tensor(?rank)) {
+        const m = max reduce t.data;
+        var y = t;
+        y.data -= m;
+        foreach i in y.data.domain {
+            y.data[i] = Math.exp(y.data[i]);
+        }
+        const sum = + reduce y.data;
+        y.data /= sum;
+        return y;
+    }
+
+    proc crossEntropyLoss(p: Tensor(1),q: Tensor(1)) {
+        var sum = 0.0;
+        forall (a,b) in zip(p,q) with (+ reduce sum){
+            sum += a * Math.log(b);
+        }
+        return -sum;
+    }
+
+    proc crossEntropyDelta(p: Tensor(1),q: Tensor(1)) {
+        return p - q;
+    }
+
+    iter convolutionRegions(dom: domain(2,int), k: int, stride: int) {
+        const (h,w) = correlateShape((k,k),dom.shape,stride,padding=0);
+        for (i,j) in {0..#h,0..#w} {
+            yield dom[i * stride..#k,j * stride..#k];
+        }
+    }
+
+    proc kernelGradient(X: Tensor(2), dY: Tensor(2),kernelSize: int,stride: int = 1) {
+        const k = kernelSize;
+        const (outH,outW) = correlateShape((k,k),X.shape,stride,padding=0);
+
+        var data: [0..#k,0..#k] real;
+        const regions = convolutionRegions(X.domain,k,stride);
+        forall (region,i) in zip(regions,0..#dY.domain.size) with (+ reduce data) {
+            data += X[region] * dY[indexInShape(dY.shape,i)];
+        }
+        var dK = new Tensor(2,real);
+        dK.reshapeDomain({0..#k,0..#k});
+        dK.data = data;
+        return dK;
+    }
+
+
+
+
+
+
+    class IainsRNG {
```

Comment:
It looks like this is no longer being used? If so, I think it's safe to remove it. Otherwise, I'd be curious if there was a limitation with the Random modules rng that you were getting around here?

Reply:
No, this was just for testing.